CommunityToolkit.Mvvm/Collections/ReadOnlyObservableGroupedCollection{TKey,TElement}.cs

```diff
@@ -0,0 +1,191 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Collections.Specialized;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace CommunityToolkit.Mvvm.Collections;
+
+/// <summary>
+/// A read-only list of groups.
+/// </summary>
+/// <typeparam name="TKey">The type of the group keys.</typeparam>
+/// <typeparam name="TElement">The type of elements in the collection.</typeparam>
+public sealed class ReadOnlyObservableGroupedCollection<TKey, TElement> : ReadOnlyObservableCollection<ReadOnlyObservableGroup<TKey, TElement>>, ILookup<TKey, TElement>
+    where TKey : notnull
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ReadOnlyObservableGroupedCollection{TKey, TValue}"/> class.
+    /// </summary>
+    /// <param name="collection">The source collection to wrap.</param>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="collection"/> is <see langword="null"/>.</exception>
+    public ReadOnlyObservableGroupedCollection(ObservableCollection<ObservableGroup<TKey, TElement>> collection)
+        : base(new ObservableCollection<ReadOnlyObservableGroup<TKey, TElement>>(collection?.Select(static g => new ReadOnlyObservableGroup<TKey, TElement>(g))!))
+    {
+        collection!.CollectionChanged += OnSourceCollectionChanged;
```

Comment:
Correct, us adding a `?` makes it think that it may in fact be null. And the compiler can't know that a null value will case the base constructor to throw, so it has to assume this line is reachable and emit that diagnostic, which we suppress.